### PR TITLE
Add protocol flag

### PR DESCRIPTION
**Pull request by juanli16**
KKRTPSI is now available as a PSI protocol for match-cli. Protocol negotiation is also implemented. Both sender and receiver have a list of preferred protocols in order of desirability. The sender sends their entire list of protocols to the receiver who compares with their own list. They search for an intersection, prioritizing their own protocol desirability. If there is an intersection, they respond with that as their decision and the sender views this as a successful negotiation. If there is no intersection, the receiver sends the default protocol (DHPSI) and the sender views this as an unsuccessful negotiation but will still proceed.

For now, the user can only input a single desired protocol in the CLI so their preferred protocol list contains a single element. In the future, multiple inputs could be allowed as the protocol negotiation already supports that functionality.

### CODE REVIEW

File: pkg/match/psi.go [hunk 1]
```diff
@@ -27,13 +27,31 @@ func Send(ctx context.Context, endpoint string, creds *tls.Config, n int64, in <
 	log.Info().Msgf("connected to partner")
 
 	// protocol negotiation step
-	protocol, err := header.NegotiateSenderProtocol(c)
+	protocolStrings := make([]string, len(preferredProtocols))
+	for i, p := range preferredProtocols {
+		protocolStrings[i] = p.String()
+	}
+	zerolog.Ctx(ctx).Info().Msgf("negotiating protocol: %v", protocolStrings)
```
Comment (juanli16) [inline pkg/match/psi.go:34]:
```suggestion
	zerolog.Ctx(ctx).Info().Msgf("negotiating protocol: %v", preferredProtocols)
```

Comment (juanli16) [reply pkg/match/psi.go:34]: String method should already take care of the formatting.

**Review by juanli16 — COMMENTED** [review]
Minor comment but LGTM!

File: pkg/match/psi.go (hunk 1 above)
Comment (xcapaldi) [reply pkg/match/psi.go]: You're right

**Comment by ratelle** [comment]
Why did we choose to still proceed if the protocol request of the sender can't be fulfilled by the receiver. Shouldn't the sender fail? I think that would be the expectation of a cli user.

File: pkg/cli/match.go [hunk 2]
```diff
@@ -39,6 +40,7 @@ type (
 		RunTimeout  time.Duration `default:"30m" help:"Timeout for the match operation"`
 		MatchID     string        `arg:"" required:"" help:"ID of the match"`
 		File        *os.File      `arg:"" required:"" help:"File to match"`
+		Protocol    string        `default:"dhpsi" enum:"bpsi,npsi,kkrtpsi,dhpsi" help:"Preferred PSI protocol"`
```
Comment (ratelle) [inline pkg/cli/match.go:43]: Match cli should only support dhpsi and kkrt

**Review by ratelle — CHANGES_REQUESTED** [review]
As mentioned, no fallback to dhpsi and no npsi and bpsi in the options.

**Comment by xcapaldi** [comment]
> Why did we choose to still proceed if the protocol request of the sender can't be fulfilled by the receiver. Shouldn't the sender fail? I think that would be the expectation of a cli user.

Okay the negotiation procedure is updated so that the whole match attempt will fail if the sender requests an unsupported protocol

**Review by ratelle — APPROVED** [review]

**Pull request merged**